### DISCUSSION TITLE
Car Colours / Handling

### DISCUSSION THREAD

**Discussion by djyt** [djyt/cannonball#91] [Ideas]
It's arguably rather trivial, but I've implemented the ability to swap colour palettes for the Ferrari. So you can have a different coloured car.  (It was actually a pain in the ass, but it's done now for the next release!)

In the long-term, this might be handy if I ever get onto crazy features like ghost-racing and/or two player support.

In the short-term I'm curious as to whether there are strong opinions on how we use this.

Option 1: Don't try to gamify this at all. Just let you choose a colour in the menu. It's purely cosmetic.
Option 2: Attach car colours to the new car handling options I'm implementing (grippy tyres, faster top speed, ability to drive off-road, ability to bump into other cars with less penalty). This is the 3DS/Switch approach.
Option 3: Take this a step further and unlock the above when you complete a route.
Option 4: Something else?!

On the one hand, as a hacker type of person, I dislike having things locked away. I want all content unlocked immediately. On the other hand, as a gamer, it's quite nice to unlock new content and gives me a reason to play and complete levels. 

![unnamed](https://user-images.githubusercontent.com/2414449/115777814-ab16cd00-a3ad-11eb-82f7-dd0ceb72580b.png)
Disclaimer: Palette shown is a quick test by me. Colours may not represent final product. Blah.

**Comment by yepyepyep4711** [comment]
Crazy ideas are good. Multiplayer over Netplay would be awesome :)

To stay on topic, this is quite philosophical :) Do you want to emphasise on showing all that is technically possible? Or is your focus on creating a fun experience? Personally, I'd gamify everything. If the unlockables are set in the config.xml anyway, those would want everything right now can unlock them there. And feel like hackers in the process ;)

**Reply by djyt** [reply]
Yeah, so to back up for a bit there were a few 'goals' with CannonBall:

1/ An education / documentation exercise in understanding how the game worked at source code level. Emulation is a 'black box' in many respects, so doesn't provide this visibility.

2/ Make respectful (and optional) enhancements that make the overall experience better. This included 60fps, widescreen, fixing bugs and adding features. (Bear in mind when I started the project there was no 3DS or Switch release and Outrun hadn't been released commercially for a while)

Part 1 is mostly complete, pending me making source code clarity improvements, which run the risk of introducing bugs and pleasing few other than myself.

Part 2 is unending. However, this did get me thinking about having some sort of Achievements/Challenge system. These could be linked with unlocks. Examples could include:

- Completing a particular route
- Overtaking a certain number of cars within a period of time
- Completing the game without crashing into scenery / other cars
- Completing the game / driving for a distance without driving off-road

I can think of more creative ones, but they potentially involve implementing new features in themselves...

**Reply by yepyepyep4711** [reply]
This all sounds good, truly. You might want to look into how Retroachievements work, though. No sense in reinventing the wheel. Actually, now that I think of it, Retroachievements (and online leaderboards), Netplay, joystick handling, all of these could be handled in Retroarch (joysticks already are), meaning you'd spare on development time. [lr-cannonball](https://github.com/libretro/cannonball) already exists, but it is not up-to-date with your code.

**Reply by djyt** [reply]
Yep I'll have a look into the achievements you mention. Chances are the next version will just have the technology and colours, and then I'll work on achievements at a later date. I prefer to release in small slices - more chance something will actually happen! 

I'm not too fussed about spending time on a libretro integration myself, but am happy if others want to maintain that library and should have a conversation at some point as to whether there's anything I can do to make life easier for whoever does.  

I can totally understand its ease of use is a positive thing for many people. But I'd rather focus on the standalone build.

**Reply by yepyepyep4711** [reply]
Î understand all your points on small slices and standalone development. Looking back, t seems to me the options in your original post are not options at all, but more the order in which you intend to do things ;)

**Reply by djyt** [reply]
Yes maybe - although if everyone shouted "we don't care about unlocks", I would have probably dropped the idea! :)
